build/gen_stub.php

```diff
@@ -2440,7 +2440,145 @@ function generateClassSynopses(array $classMap): array {
  */
 function replaceClassSynopses(string $targetDirectory, array $classMap): array
 {
-    throw new Exception("Not yet implemented!");
+    $classSynopses = [];
+
+    $it = new RecursiveIteratorIterator(
+        new RecursiveDirectoryIterator($targetDirectory),
+        RecursiveIteratorIterator::LEAVES_ONLY
+    );
+
+    foreach ($it as $file) {
+        $pathName = $file->getPathName();
+        if (!preg_match('/\.xml$/i', $pathName)) {
+            continue;
+        }
+
+        $xml = file_get_contents($pathName);
+        if ($xml === false) {
+            continue;
+        }
+
+        if (stripos($xml, "<classsynopsis") === false) {
+            continue;
+        }
+
+        $replacedXml = preg_replace(
+            [
+                "/&([A-Za-z0-9._{}%-]+?;)/",
+                "/<(\/)*xi:([A-Za-z]+?)/"
+            ],
+            [
+                "REPLACED-ENTITY-$1",
+                "<$1XI$2",
+            ],
+            $xml
+        );
+
+        $doc = new DOMDocument();
+        $doc->formatOutput = false;
+        $doc->preserveWhiteSpace = true;
+        $doc->validateOnParse = true;
+        $success = $doc->loadXML($replacedXml);
+        if (!$success) {
+            echo "Failed opening $pathName\n";
+            continue;
+        }
+
+        $docComparator = new DOMDocument();
+        $docComparator->preserveWhiteSpace = false;
+        $docComparator->formatOutput = true;
+
+        $classSynopsisElements = [];
+        foreach ($doc->getElementsByTagName("classsynopsis") as $element) {
+            $classSynopsisElements[] = $element;
+        }
+
+        foreach ($classSynopsisElements as $classSynopsis) {
+            if (!$classSynopsis instanceof DOMElement) {
+                continue;
+            }
+
+            $firstChild = $classSynopsis->firstElementChild;
+            if ($firstChild === null) {
+                continue;
+            }
+            $firstChild = $firstChild->firstElementChild;
+            if ($firstChild === null) {
+                continue;
+            }
+            $className = $firstChild->textContent;
+            if (!isset($classMap[$className])) {
+                continue;
+            }
+            $classInfo = $classMap[$className];
+
+            $newClassSynopsis = $classInfo->getClassSynopsisElement($doc, $classMap);
+            if ($newClassSynopsis === null) {
+                continue;
+            }
+
+            // Check if there is any change - short circuit if there is not any.
+
+            $xml1 = $doc->saveXML($classSynopsis);
+            $xml1 = preg_replace(
+                [
+                    "/&([A-Za-z0-9._{}%-]+?;)/",
+                    "/<(\/)*xi:([A-Za-z]+?)/",
+                ],
+                [
+                    "REPLACED-ENTITY-$1",
+                    "<$1XI$2",
+                ],
+                $xml1
+            );
+            $docComparator->loadXML($xml1);
+            $xml1 = $docComparator->saveXML();
```

Comment:
I don't really get what this is doing ... why do we need to go through a loadXML+saveXML round trip here?

Reply:
Finally I found it out again why it's needed for class and method synopses: `$docComparator` doesn't preserve whitespaces (while `$doc` does), so we don't replace the class synopsis if only some whitespace changed.

Maybe I should set `$doc->preserveWhitespace` to `false` before I compare the contents, and set it `true` afterwards?

Reply:
I think just having a separate function `areDocsEqualApartFromWhitespace()` or so would make it obvious what's going on.

Reply:
Though if preserveWhitespace only affects the saveXML step, then setting that before the saveXML calls also makes sense to me.

Reply:
> Though if preserveWhitespace only affects the saveXML step, then setting that before the saveXML calls also makes sense to me.

No, unfortunately it seems that it doesn't only affect the saveXML step, so I couldn't make my idea above work (and I'm sure that I tried a lot of things when I worked on the method synopsis replacement, but this was the only one which worked).

> I think just having a separate function areDocsEqualApartFromWhitespace() or so would make it obvious what's going on.

There's a comment a few lines above: `// Check if there is any change - short circuit if there is not any.` The reason why I didn't want to explode the functionality of the function into smaller pieces is that somehow I'm a bit reluctant to add global utility functions which should otherwise be private methods. But that's probably me :) So if that's not a problem for you then I'll yet add `areDocsEqualApartFromWhitespace()`.

Reply:
I don't mind extra utility functions, though I guess the getReplacedClassSynopsisXml() call here would mean that this can't be shared with the method synopsis code, that does a different replacement...

Reply:
The only difference between them that class synopses have the`<xi:include>` tags which should be replaced as well. Since method synopses don't contain such tag, we could extract this functionality in fact. I'll commit it in one sec

Reply:
I've just pushed it, and you were right, the code is much better this way :)